gw_eccentricity/eccDefinition.py

```diff
@@ -1285,33 +1442,217 @@ def get_good_extrema(self, pericenters, apocenters,
+
+        This function ensures that the rational fit uses the optimal degree for
+        the numerator and denominator. It first checks for nonmonotonicity in
+        the fit's derivative and lowers the degrees if necessary. Only if the
+        fit for the initial degree was monotonic (no reduction of degree
+        needed), it attempts to increase the degrees to find a higher-degree
+        fit. If increasing the degrees causes nonmonotonicity, it reverts to
+        the previous valid monotonic fit.
+
+        The initial degrees for the rational fit can be specified through
+        `special_interp_kwargs_for_extrema` in `extra_kwargs` with the key
+        "rational_fit". (See `special_interp_kwargs_for_extrema` under
+        `extra_kwargs` for more details) Default values are provided by
+        `get_default_rational_fit_kwargs`, where both degrees are set to
+        `None`. If both degrees remain `None`, appropriate starting values are
+        determined using `self.get_approximate_degree_for_rational_fit`.
         """
-        return get_interpolant(oldX, oldY, allowExtrapolation, interpolator,
-                               spline_kwargs=self.spline_kwargs,
-                               check_kwargs=False)
+        # make sure that data_name is not None. A data_name is needed to
+        # update the optimal values of the numerator and denominator degrees
+        # used to build the final rational fit
+        if data_name is None:
+            raise Exception(
+                "Please provide a data_name for which to build a "
+                "rational fit. `data_name` can not be None. For example, "
+                "it can be 'apocenters', 'pericenters' or 'omega_gw_average'.")
+        # Set initial degrees for the rational fit if not already specified. If
+        # either `num_degree` or `denom_degree` is specified, use that value to
+        # set the other. If both are unspecified, we use the following
+        # procedure to set appropriate values for `num_degree` and
+        # `denom_degree`.
+        
+        # The optimal degrees for the numerator and denominator can vary with
+        # the eccentricity and duration of the waveform. To estimate these
+        # values, we first count the approximate number of orbits, then use a
+        # predefined mapping from orbit count to optimal degrees, provided by
+        # `get_approximate_degree_for_rational_fit`. 
+        
+        # While this provides a good initial estimate, it may not be optimal,
+        # as the ideal degree can change based on the eccentricity which we do
+        # not know apriori. The true optimal values are determined through
+        # further iterations.
+        if self.special_interp_kwargs_for_extrema["num_degree"] is None \
+            and self.special_interp_kwargs_for_extrema["denom_degree"] is None:
+            self.special_interp_kwargs_for_extrema["num_degree"], \
```

Comment:
Mention in a comment what this code is doing: If only one of num_degree/denom_degree is given, it sets the other one to match it.